docs: Introduce spec overview and template

powersj opened the pull request:
## Summary
Introduces the spec background and process for Telegraf.

## Checklist
<!-- Mandatory
Please confirm the following by replacing the space with an "x" between the []:
-->

- [x] No AI generated code was used in this PR

srebhan commented on docs/specs/overview.md:
```diff
@@ -0,0 +1,57 @@
+
+## Objective
+
+Define and layout the Telegraf spec process.
```
```suggestion
Define and layout the Telegraf specification process.
```

srebhan commented on docs/specs/overview.md:
```diff
@@ -0,0 +1,57 @@
+# Telegraf Spec Overview
```
```suggestion
# Telegraf Specification Overview
```

srebhan commented on docs/specs/overview.md:
```diff
@@ -0,0 +1,57 @@
+decent understanding of the objective, the steps required, and most of the
+general design decisions.
+
+The specs can then live in the Telegraf repo to share and involve with the
+community the planning process that goes into larger changes and have a public
+historical record for changes.
```
```suggestion
The specs can then live in the Telegraf repository to share and involve with the
community in the process of planning larger changes or new features. The specs
also serve as a public historical record for changes.
```

srebhan reviewed (commented):
Thanks @powersj! Some small wording suggestions...

I wonder if we should add more meta-information to the spec document e.g.
- defining a naming scheme for the specs to be able to reference those reliably
- parts of Telegraf this applies to e.g. `inputs`, `agent`... (e.g. using keywords)
- having a short abstract of what the spec is about, i.e. making the objective a bit longer (usually scientific papers limit the abstract to 150 words which seems to be a good thing)

What do you think?

Hipska commented on docs/specs/overview.md:
Maybe rename this to README.md so GitHub directly opens this one when opening the specs dir?

powersj replied:
agreed - done

srebhan commented:
@powersj while I do like expressive naming I think having a unique number is also important, so how about a naming scheme: `<prefix>-<three-digit-ID>-<name>.md`, where `prefix` is e.g. `TSD` (telegraf spec document) or something like this, then a consecutive ID and then a free name what the spec is about. This way you can reference the spec via `TSD-011` and we know where to find it!?

I would love to see a `Subject` in the spec like

```markdown
# Startup retry configuration

## Objective

Allow configuration of plugin behavior on startup errors and unify settings.

## Subject

inputs, processors, aggregators, outputs, startup, error behavior

## Overview

...
```

Where the `Subject` section has a handful of keywords defining what is touched/targeted by the spec. What do you think?

powersj commented:
> This way you can reference the spec via TSD-011 and we know where to find it!?

I am ok with this and have updated the doc. I do personally dislike referring to specs by number as it doesn't give context, consider "the write ahead spec" vs "TSD-10". I do not expect us to have 100s of these, I have been here 2 and half years and we have less than 25 specs.

> Where the Subject section has a handful of keywords defining what is touched/targeted by the spec. What do you think?

I am not opposed to suggesting this, but what value does it provide? 

Keep in mind that every section you request is another barrier to entry that slows down development of specs or reduces the likelihood of community additions. Therefore, I want to be very clear as to why we add something and why we are expecting it.

srebhan commented:
Well numbering has the advantage of preserving order e.g. `TSD-010-write-ahead-log.md` and `TSD-017-write-ahead-log.md` provides a clear order of what is the latest take.

For the keyworks, I think it is just a list of 1 to X words making clear what the spec touches, like for the WAL, it should be "outputs" making clear it does not work on the input side... But I'm fine with leaving this out.

powersj commented:
> Well numbering has the advantage of preserving order e.g. TSD-010-write-ahead-log.md and TSD-017-write-ahead-log.md provides a clear order of what is the latest take.

ok thanks for this!

> For the keyworks, I think it is just a list of 1 to X words making clear what the spec touches, like for the WAL, it should be "outputs" making clear it does not work on the input side... But I'm fine with leaving this out.

Thanks for the background. I've added this to the optional sections and updated the template.

Hipska commented on docs/specs/template.md:
Maybe rename this one as `tsd-000-template.md`?
